refactor: remove deprecated lifecycle methods in Table

afc163 opened the pull request:
<!--
First of all, thank you for your contribution! 😄

New feature please send pull request to feature branch, and rest to master branch.
Pull request will be merged after one of collaborators approve.
Please makes sure that these form are filled before submitting your pull request, thank you!

[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]
-->

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

<!--
1. Describe the source of requirement, like related issue link.
-->
https://github.com/ant-design/ant-design/issues/9792

#18250

### 💡 Background and solution

<!--
1. Describe the problem and the scenario.
2. GIF or snapshot should be provided if includes UI/interactive modification.
3. How to fix the problem, and list final API implementation and usage sample if that is an new feature.
-->
Updated to `react` v16.9 and it started to log warnings about the unsafe lifecycle methods.

<details>
<summary>📸 Screenshot</summary>

![image](https://user-images.githubusercontent.com/20532589/62952503-0f694700-bdec-11e9-9d1c-042dc590b484.png)

</details>

I ran `npx react-codemod rename-unsafe-lifecycles`, following the React [blog post](https://reactjs.org/blog/2019/08/08/react-v16.9.0.html#renaming-unsafe-lifecycle-methods).

### 📝 Changelog

No more warnings about `componentWillReceiveProps` from `Table` and `SelectionCheckboxAll` components. 🙅‍♂ 

<!--
Describe changes from user-side, and list all potential break changes or other risks.
--->

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     -      |
| 🇨🇳 Chinese |     -      |

### ☑️ Self Check before Merge

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

kristofdombi committed:
refactor(Table): rename cWRP to UNSAFE_cWRP

kristofdombi committed:
fix(Table, SelectionCheckboxAll): add eslint comment

kristofdombi committed:
refactor(Table): move createComponents outside of class, eliminate cWRP

kristofdombi committed:
fix(Table): rm commented line

kristofdombi committed:
fix(Table): move rowSelection logic to getDerivedStateFromProps

kristofdombi committed:
refactor(Table): move columns to state to set in getDerivedState

kristofdombi committed:
refactor(Table): move filter updating to getDerivedStateFromProps

kristofdombi committed:
fix(Table): prevent reusing row from previous render

kristofdombi committed:
refactor(Table): make row work like a class property

kristofdombi committed:
fix(Table): correct lint errors

orzyyyy commented on components/table/Table.tsx:
```diff
@@ -87,7 +90,65 @@ const defaultPagination = {
+  const prevBodyRow = prevComponents && prevComponents.body && prevComponents.body.row;
+  if (isCalledFromConstructor) {
+    // 'tableRow' needs to act like a class property
+    tableRow = undefined;
```
I don't quite understand why adding this, what I think `tableRow` is `undefined` in constructor already. Any other where the `tableRow` had been used?

afc163 replied:
@kristof0425

afc163 replied:
https://github.com/ant-design/ant-design/pull/18953/commits/a5ad47b4870af0d510798ba0193fdcf9eb102265

shaodahong commented on components/table/Table.tsx:
```diff
@@ -148,89 +270,38 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
-
-    if (this.getSortOrderColumns(this.columns).length > 0) {
-      const sortState = this.getSortStateFromColumns(this.columns);
+  componentDidUpdate() {
```
componentDidUpdate lifecycle maybe not need. getDerivedStateFromProps do same thing

orzyyyy replied:
Maybe `getSnapshotBeforeUpdate` is more suitable

yoyo837 commented on components/table/Table.tsx:
```diff
@@ -124,22 +185,83 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
+        ...nextProps.pagination,
+      };
+      newPagination.current = newPagination.current || 1;
+      newPagination.pageSize = newPagination.pageSize || 10;
```
pageSizeOptions[0] ?

afc163 commented:
/rebase

afc163 commented:
/rebase

afc163 commented:
/rebase

afc163 committed:
:lipstick: improve code style

afc163 committed:
:hammer: react-lifecycles-compat for Table

afc163 committed:
:zap: optimize Table components compare logic

orzyyyy commented:
看上去只是重构，为啥不发到 `master` 去？

afc163 committed:
:bug: fix React.createElement undefined

afc163 commented:
@orzyyyy

比较大的重构，稳妥起见放到 minor 上。

afc163 committed:
:white_check_mark: more test cases

afc163 committed:
:white_check_mark: increase Table test coverage

afc163 committed:
:bug: Fix Table filter submenu checkbox margin

https://user-images.githubusercontent.com/507615/65814464-a5c6c000-e214-11e9-8085-2c6fb1f5c6f2.png

afc163 committed:
:lipstick: fix lint

afc163 committed:
:white_check_mark: increase Table test coverage

afc163 commented:
谁在帮忙审核一下，给一个 approve。

patrickomeara commented:
@kristof0425 This is a great PR, well written. I've just updated from `v3.23.6` but wherever I am using `<Table />` with `TableProps` now wants `store`, `checkboxPropsCache`, and `setCheckboxPropsCache`. 

Was there a reason these weren't made optional?

javierfz1980 commented:
I have the same issue reported by @patrickomeara <Table /> now requires store, checkboxPropsCache, setCheckboxPropsCache as mandatory...

suraneti commented:
I got warning componentWillReceiveProps from using the `notification` component on the latest antd.
